compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -498,21 +500,98 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                     self.suggest_cloning(err, ty, expr, None);
                 }
             }
+
+            self.suggest_ref_for_dbg_args(expr, span, move_span, err);
+
             if let Some(pat) = finder.pat {
-                *in_pattern = true;
-                let mut sugg = vec![(pat.span.shrink_to_lo(), "ref ".to_string())];
-                if let Some(pat) = finder.parent_pat {
-                    sugg.insert(0, (pat.span.shrink_to_lo(), "ref ".to_string()));
-                }
-                err.multipart_suggestion_verbose(
-                    "borrow this binding in the pattern to avoid moving the value",
-                    sugg,
-                    Applicability::MachineApplicable,
+                // FIXME: any better way to check this?
+                let from_std = self.infcx.tcx.sess.opts.real_rust_source_base_dir.clone().map_or(
+                    false,
+                    |root| {
+                        let file_path =
+                            match self.infcx.tcx.sess.source_map().span_to_filename(move_span) {
+                                FileName::Real(name) => {
+                                    name.clone().into_local_path().unwrap_or_default()
+                                }
+                                other => PathBuf::from(other.prefer_local().to_string()),
+                            };
+                        file_path.starts_with(&root.join("library/std/"))
+                    },
                 );
+                // it's useless to suggest inserting `ref` when the span comes from std library
+                // anyway, user can not modify std library in most cases, so let's keep it quite?
+                if !from_std {
+                    *in_pattern = true;
+                    let mut sugg = vec![(pat.span.shrink_to_lo(), "ref ".to_string())];
+                    if let Some(pat) = finder.parent_pat {
+                        sugg.insert(0, (pat.span.shrink_to_lo(), "ref ".to_string()));
+                    }
+                    err.multipart_suggestion_verbose(
+                        "borrow this binding in the pattern to avoid moving the value",
+                        sugg,
+                        Applicability::MachineApplicable,
+                    );
+                }
             }
         }
     }
 
+    // for dbg!(x) which may take onwership, suggest dbg!(&x) instead
+    // but here we actually does not checking the macro name is `dbg!`
+    // so that we may extend the scope a bit larger to cover more cases
+    fn suggest_ref_for_dbg_args(
+        &self,
+        body: &hir::Expr<'_>,
+        span: Option<Span>,
+        move_span: Span,
+        err: &mut Diag<'tcx>,
+    ) {
+        // only suggest for macro
+        if move_span.source_callsite() == move_span {
+            return;
+        }
```

Comment:
why? How do the diagnostics look without macros and could they improved by this, too?

Reply:
If we don't limit only for macro, it will also add suggestion for this code:

```rust
fn s() -> String {
    let a = String::new();
    let _b = match a {
        tmp => {
            eprintln!("dbg: {}", tmp);
            tmp
        }
    };
    return a;
}

```

with diagnostic:

```console
 --> ./t/ha.rs:15:12
   |
7  |     let a = String::new();
   |         - move occurs because `a` has type `String`, which does not implement the `Copy` trait
...
10 |         tmp => {
   |         --- value moved here
...
15 |     return a; //~ ERROR use of moved value:
   |            ^ value used here after move
   |
help: consider borrowing instead of transferring ownership
   |
9  |     let _b = match &a {
   |                    +
help: borrow this binding in the pattern to avoid moving the value
   |
10 |         ref tmp => {
   |         +++
```

Seems it's also a good one.

I will add it into testcase.